src/vision/deliverables/async_fn/async_drop.md

```diff
@@ -0,0 +1,115 @@
+
+### Explicit async drop
+
+We should have a `std::mem::async_drop` analogous to `std::mem::drop`:
+
+```rust
+async fn async_drop<T>(x: T) { }
+```
+
+### Implicit await points
+
+When you run async drop glue, there is an implicit await point. Consider this example:
+
+```rust
+async fn foo(dbc: DatabaseConnection) -> io::Result<()> {
+    let data = socket().read().await?;
+    dbc.write(data).await?;
+}
+```
+
+Here, presuming that `DatabaseConnection` implements `AsyncDrop`, there are actually a number of async drops occurring:
+
+```rust
+async fn foo(dbc: DatabaseConnection) -> io::Result<()> {
+    let data = match socket().read().await {
+        Ok(v) => v,
+        Err(e) => {
+            std::mem::async_drop(dbc).await;
+            return e;
+        }
+    };
+    let () = match dbc.write(data).await? {
+        Ok(()) => (),
+        Err(e) => {
+            std::mem::async_drop(dbc).await;
+            return e;
+        }
+    };
+    std::mem::async_drop(dbc).await;
+}
+```
+
+As this example shows, there are important ergonomic benefits here to implicit async drop, and it also ensures that async and sync code work in analogous ways. However, implicit await points can be a hazard for some applications, where it is important to identify all await points explicitly (for example, authors of embedded applications use await points to reason about what values will be stored in the resulting future vs the stack of the poll function). To further complicate things, async-drop doesn't only execute at the end of a block or an "abrupt" expression like `?`: async-drop can also execute at the end of every statement, given temporary values.
+
+The best solution here is unclear. We could have an "allow-by-default" lint encouraging explicit use of `async_drop`, but as the code above shows, the result may be highly unergonomic (also, imagine how it looks as the number of variables requiring async-drop grows).
+
+Another option is to target the problem from another angle, for example by adding lints to identify when large values are stored in a future or on the stack, or to allow developers to tag local variables that they expect to be stored on the stack, and have the compiler warn them if this turns out to not be true.
+
+### Preventing sync drop
+
+It is easy enough to make async-drop be used, but it is currently not possible to prevent sync drop, even from within an async setting. Consider an example such as the following:
+
+```rust
+async fn foo(dbc: DatabaseConnection) -> io::Result<()> {
+    drop(dbc);
+}
+```
+
+The compiler could however lint against invoking (or defining!) synchronous functions that take ownership of values whose types implement `AsyncDrop`. This would catch code like the case above. We may have to tune the lint to avoid false warnings. Note that it is important to lint *both* invocation *and* definition sites because the synchronous function may be generic (like `drop`, in fact).
```

Comment:
What do you mean by linting definition sites? Something that takes `impl Async` and drops it?